// skynet-js/dist/cjs/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkynetClient = void 0;
const axios_1 = __importDefault(require("axios"));
const skynet_mysky_utils_1 = require("skynet-mysky-utils");
const upload_1 = require("./upload");
const download_1 = require("./download");
// These imports are deprecated but they are needed to export the v1 File
// methods, which we are keeping so as not to break compatibility.
const file_1 = require("./file");
const pin_1 = require("./pin");
const registry_1 = require("./registry");
const revision_cache_1 = require("./revision_cache");
// These imports are deprecated but they are needed to export the v1 SkyDB
// methods, which we are keeping so as not to break compatibility.
const skydb_1 = require("./skydb");
const skydb_v2_1 = require("./skydb_v2");
const url_1 = require("./utils/url");
const mysky_1 = require("./mysky");
const utils_1 = require("./mysky/utils");
const request_1 = require("./request");
// Add a response interceptor so that we always return an error of type
// `ExecuteResponseError`.
axios_1.default.interceptors.response.use(function (response) {
    // Any status code that lie within the range of 2xx cause this function to trigger.
    // Do something with response data.
    return response;
}, function (error) {
    // Any status codes that falls outside the range of 2xx cause this function to trigger
    // Do something with response error.
    return Promise.reject(request_1.ExecuteRequestError.From(error));
});
/**
 * The Skynet Client which can be used to access Skynet.
 */
class SkynetClient {
    /**
     * The Skynet Client which can be used to access Skynet.
     *
     * @class
     * @param [initialPortalUrl] The initial portal URL to use to access Skynet, if specified. A request will be made to this URL to get the actual portal URL. To use the default portal while passing custom options, pass "".
     * @param [customOptions] Configuration for the client.
     */
    constructor(initialPortalUrl = "", customOptions = {}) {
        // Set methods (defined in other files).
        // Upload
        this.uploadFile = upload_1.uploadFile;
        this.uploadSmallFile = upload_1.uploadSmallFile;
        this.uploadSmallFileRequest = upload_1.uploadSmallFileRequest;
        this.uploadLargeFile = upload_1.uploadLargeFile;
        this.uploadLargeFileRequest = upload_1.uploadLargeFileRequest;
        this.uploadDirectory = upload_1.uploadDirectory;
        this.uploadDirectoryRequest = upload_1.uploadDirectoryRequest;
        // Download
        this.downloadFile = download_1.downloadFile;
        this.downloadFileHns = download_1.downloadFileHns;
        this.getSkylinkUrl = download_1.getSkylinkUrl;
        this.getHnsUrl = download_1.getHnsUrl;
        this.getHnsresUrl = download_1.getHnsresUrl;
        this.getMetadata = download_1.getMetadata;
        this.getFileContent = download_1.getFileContent;
        this.getFileContentRequest = download_1.getFileContentRequest;
        this.getFileContentHns = download_1.getFileContentHns;
        this.openFile = download_1.openFile;
        this.openFileHns = download_1.openFileHns;
        this.resolveHns = download_1.resolveHns;
        // Pin
        this.pinSkylink = pin_1.pinSkylink;
        // MySky
        this.extractDomain = utils_1.extractDomain;
        this.getFullDomainUrl = utils_1.getFullDomainUrl;
        this.loadMySky = mysky_1.loadMySky;
        // File API
        this.file = {
            getJSON: file_1.getJSON.bind(this),
            getEntryData: file_1.getEntryData.bind(this),
            getEntryLink: file_1.getEntryLink.bind(this),
            getJSONEncrypted: file_1.getJSONEncrypted.bind(this),
        };
        // SkyDB
        // v1 (deprecated)
        this.db = {
            getJSON: skydb_1.getJSON.bind(this),
            setJSON: skydb_1.setJSON.bind(this),
            deleteJSON: skydb_1.deleteJSON.bind(this),
            getRawBytes: skydb_1.getRawBytes.bind(this),
            setDataLink: skydb_1.setDataLink.bind(this),
            getEntryData: skydb_1.getEntryData.bind(this),
            setEntryData: skydb_1.setEntryData.bind(this),
            deleteEntryData: skydb_1.deleteEntryData.bind(this),
        };
        // v2
        this.dbV2 = {
            getJSON: skydb_v2_1.getJSON.bind(this),
            setJSON: skydb_v2_1.setJSON.bind(this),
            deleteJSON: skydb_v2_1.deleteJSON.bind(this),
            getRawBytes: skydb_v2_1.getRawBytes.bind(this),
            setDataLink: skydb_v2_1.setDataLink.bind(this),
            getEntryData: skydb_v2_1.getEntryData.bind(this),
            setEntryData: skydb_v2_1.setEntryData.bind(this),
            deleteEntryData: skydb_v2_1.deleteEntryData.bind(this),
            // Holds the cached revision numbers, protected by mutexes to prevent
            // concurrent access.
            revisionNumberCache: new revision_cache_1.RevisionNumberCache(),
        };
        // Registry
        this.registry = {
            getEntry: registry_1.getEntry.bind(this),
            getEntryUrl: registry_1.getEntryUrl.bind(this),
            getEntryLink: registry_1.getEntryLinkAsync.bind(this),
            setEntry: registry_1.setEntry.bind(this),
            postSignedEntry: registry_1.postSignedEntry.bind(this),
        };
        if (initialPortalUrl === "") {
            // Portal was not given, use the default portal URL. We'll still make a request for the resolved portal URL.
            initialPortalUrl = (0, url_1.defaultPortalUrl)();
        }
        else {
            // Portal was given, don't make the request for the resolved portal URL.
            this.customPortalUrl = (0, skynet_mysky_utils_1.ensureUrl)(initialPortalUrl);
        }
        this.initialPortalUrl = initialPortalUrl;
        this.customOptions = customOptions;
    }
    /* istanbul ignore next */
    /**
     * Make the request for the API portal URL.
     *
     * @returns - A promise that resolves when the request is complete.
     */
    async initPortalUrl() {
        if (this.customPortalUrl) {
            // Tried to make a request for the API portal URL when a custom URL was already provided.
            return;
        }
        // Try to resolve the portal URL again if it's never been called or if it
        // previously failed.
        if (!SkynetClient.resolvedPortalUrl) {
            SkynetClient.resolvedPortalUrl = this.resolvePortalUrl();
        }
        else {
            try {
                await SkynetClient.resolvedPortalUrl;
            }
            catch (e) {
                SkynetClient.resolvedPortalUrl = this.resolvePortalUrl();
            }
        }
        // Wait on the promise and throw if it fails.
        await SkynetClient.resolvedPortalUrl;
        return;
    }
    /* istanbul ignore next */
    /**
     * Returns the API portal URL. Makes the request to get it if not done so already.
     *
     * @returns - the portal URL.
     */
    async portalUrl() {
        if (this.customPortalUrl) {
            return this.customPortalUrl;
        }
        // Make the request if needed and not done so.
        await this.initPortalUrl();
        return await SkynetClient.resolvedPortalUrl; // eslint-disable-line
    }
    /**
     * Creates and executes a request.
     *
     * @param config - Configuration for the request.
     * @returns - The response from axios.
     * @throws - Will throw `ExecuteRequestError` if the request fails. This error contains the original Axios error.
     */
    async executeRequest(config) {
        const url = await (0, request_1.buildRequestUrl)(this, {
            baseUrl: config.url,
            endpointPath: config.endpointPath,
            subdomain: config.subdomain,
            extraPath: config.extraPath,
            query: config.query,
        });
        // Build headers.
        const headers = (0, request_1.buildRequestHeaders)(config.headers, config.customUserAgent, config.customCookie, config.skynetApiKey);
        const auth = config.APIKey ? { username: "", password: config.APIKey } : undefined;
        let onDownloadProgress = undefined;
        if (config.onDownloadProgress) {
            onDownloadProgress = function (event) {
                // Avoid NaN for 0-byte file.
                /* istanbul ignore next: Empty file test doesn't work yet. */
                const progress = event.total ? event.loaded / event.total : 1;
                // @ts-expect-error TS complains even though we've ensured this is defined.
                config.onDownloadProgress(progress, event);
            };
        }
        let onUploadProgress = undefined;
        if (config.onUploadProgress) {
            onUploadProgress = function (event) {
                // Avoid NaN for 0-byte file.
                /* istanbul ignore next: event.total is always 0 in Node. */
                const progress = event.total ? event.loaded / event.total : 1;
                // @ts-expect-error TS complains even though we've ensured this is defined.
                config.onUploadProgress(progress, event);
            };
        }
        // NOTE: The error type will be `ExecuteRequestError` as we set up a
        // response interceptor above.
        try {
            return await (0, axios_1.default)({
                url,
                method: config.method,
                data: config.data,
                headers,
                auth,
                onDownloadProgress,
                onUploadProgress,
                responseType: config.responseType,
                transformRequest: config.transformRequest,
                transformResponse: config.transformResponse,
                maxContentLength: Infinity,
                maxBodyLength: Infinity,
                // Allow cross-site cookies.
                withCredentials: true,
            });
        }
        catch (e) {
            if (config.loginFn && e.responseStatus === 401) {
                // Try logging in again.
                await config.loginFn();
                return await this.executeRequest(config);
            }
            else {
                throw e;
            }
        }
    }
    // ===============
    // Private Methods
    // ===============
    /**
     * Gets the current server URL for the portal. You should generally use
     * `portalUrl` instead - this method can be used for detecting whether the
     * current URL is a server URL.
     *
     * @returns - The portal server URL.
     */
    async resolvePortalServerUrl() {
        const response = await this.executeRequest({
            ...this.customOptions,
            method: "head",
            url: this.initialPortalUrl,
        });
        if (!response.headers) {
            throw new Error("Did not get 'headers' in response despite a successful request. Please try again and report this issue to the devs if it persists.");
        }
        const portalUrl = response.headers["skynet-server-api"];
        if (!portalUrl) {
            throw new Error("Could not get server portal URL for the given portal");
        }
        return portalUrl;
    }
    /**
     * Make a request to resolve the provided `initialPortalUrl`.
     *
     * @returns - The portal URL.
     */
    async resolvePortalUrl() {
        const response = await this.executeRequest({
            ...this.customOptions,
            method: "head",
            url: this.initialPortalUrl,
        });
        if (!response.headers) {
            throw new Error("Did not get 'headers' in response despite a successful request. Please try again and report this issue to the devs if it persists.");
        }
        const portalUrl = response.headers["skynet-portal-api"];
        if (!portalUrl) {
            throw new Error("Could not get portal URL for the given portal");
        }
        return portalUrl;
    }
}
exports.SkynetClient = SkynetClient;
