// skynet-js/dist/cjs/crypto.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sha512 = exports.hashRegistryEntry = exports.hashDataKey = exports.hashAll = exports.genKeyPairFromSeed = exports.genKeyPairAndSeed = exports.deriveChildSeed = exports.HASH_LENGTH = void 0;
const sjcl_1 = require("sjcl");
const blakejs_1 = require("blakejs");
const randombytes_1 = __importDefault(require("randombytes"));
const tweetnacl_1 = require("tweetnacl");
const string_1 = require("./utils/string");
const validation_1 = require("./utils/validation");
const encoding_1 = require("./utils/encoding");
exports.HASH_LENGTH = 32;
/**
 * Returns a blake2b 256bit hasher. See `NewHash` in Sia.
 *
 * @returns - blake2b 256bit hasher.
 */
function newHash() {
    return blakejs_1.blake2bInit(32, null);
}
/**
 * Derives a child seed from the given master seed and sub seed.
 *
 * @param masterSeed - The master seed to derive from.
 * @param seed - The sub seed for the derivation.
 * @returns - The child seed derived from `masterSeed` using `seed`.
 * @throws - Will throw if the inputs are not strings.
 */
function deriveChildSeed(masterSeed, seed) {
    validation_1.validateString("masterSeed", masterSeed, "parameter");
    validation_1.validateString("seed", seed, "parameter");
    return string_1.toHexString(hashAll(encoding_1.encodeUtf8String(masterSeed), encoding_1.encodeUtf8String(seed)));
}
exports.deriveChildSeed = deriveChildSeed;
/**
 * Generates a master key pair and seed.
 *
 * @param [length=64] - The number of random bytes for the seed. Note that the string seed will be converted to hex representation, making it twice this length.
 * @returns - The generated key pair and seed.
 */
function genKeyPairAndSeed(length = 64) {
    validation_1.validateNumber("length", length, "parameter");
    const seed = makeSeed(length);
    return { ...genKeyPairFromSeed(seed), seed };
}
exports.genKeyPairAndSeed = genKeyPairAndSeed;
/**
 * Generates a public and private key from a provided, secure seed.
 *
 * @param seed - A secure seed.
 * @returns - The generated key pair.
 * @throws - Will throw if the input is not a string.
 */
function genKeyPairFromSeed(seed) {
    validation_1.validateString("seed", seed, "parameter");
    // Get a 32-byte key.
    const derivedKey = sjcl_1.misc.pbkdf2(seed, "", 1000, 32 * 8);
    const derivedKeyHex = sjcl_1.codec.hex.fromBits(derivedKey);
    const { publicKey, secretKey } = tweetnacl_1.sign.keyPair.fromSeed(string_1.hexToUint8Array(derivedKeyHex));
    return { publicKey: string_1.toHexString(publicKey), privateKey: string_1.toHexString(secretKey) };
}
exports.genKeyPairFromSeed = genKeyPairFromSeed;
/**
 * Takes all given arguments and hashes them.
 *
 * @param args - Byte arrays to hash.
 * @returns - The final hash as a byte array.
 */
function hashAll(...args) {
    const hasher = newHash();
    args.forEach((arg) => blakejs_1.blake2bUpdate(hasher, arg));
    return blakejs_1.blake2bFinal(hasher);
}
exports.hashAll = hashAll;
// TODO: Is this the same as hashString?
/**
 * Hash the given data key.
 *
 * @param dataKey - Data key to hash.
 * @returns - Hash of the data key.
 */
function hashDataKey(dataKey) {
    return hashAll(encoding_1.encodeUtf8String(dataKey));
}
exports.hashDataKey = hashDataKey;
/**
 * Hashes the given registry entry.
 *
 * @param registryEntry - Registry entry to hash.
 * @param hashedDataKeyHex - Whether the data key is already hashed and in hex format. If not, we hash the data key.
 * @returns - Hash of the registry entry.
 */
function hashRegistryEntry(registryEntry, hashedDataKeyHex) {
    let dataKeyBytes;
    if (hashedDataKeyHex) {
        dataKeyBytes = string_1.hexToUint8Array(registryEntry.dataKey);
    }
    else {
        dataKeyBytes = hashDataKey(registryEntry.dataKey);
    }
    const dataBytes = encoding_1.encodePrefixedBytes(registryEntry.data);
    return hashAll(dataKeyBytes, dataBytes, encoding_1.encodeBigintAsUint64(registryEntry.revision));
}
exports.hashRegistryEntry = hashRegistryEntry;
/**
 * Hashes the given string or byte array using sha512.
 *
 * @param message - The string or byte array to hash.
 * @returns - The resulting hash.
 */
function sha512(message) {
    if (typeof message === "string") {
        return tweetnacl_1.hash(string_1.stringToUint8ArrayUtf8(message));
    }
    else {
        return tweetnacl_1.hash(message);
    }
}
exports.sha512 = sha512;
/**
 * Generates a random seed of the given length in bytes.
 *
 * @param length - Length of the seed in bytes.
 * @returns - The generated seed.
 */
function makeSeed(length) {
    // Cryptographically-secure random number generator. It should use the
    // built-in crypto.getRandomValues in the browser.
    const array = randombytes_1.default(length);
    return string_1.toHexString(array);
}
