// skynet-js/dist/cjs/mysky/encrypted_files.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.encodeEncryptedFileMetadata = exports.checkPaddedBlock = exports.padFileSize = exports.deriveEncryptedFileSeed = exports.deriveEncryptedFileTweak = exports.deriveEncryptedFileKeyEntropy = exports.encryptJSONFile = exports.decryptJSONFile = exports.ENCRYPTION_PATH_SEED_LENGTH = exports.ENCRYPTION_NONCE_LENGTH = exports.ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH = exports.ENCRYPTION_KEY_LENGTH = exports.ENCRYPTED_JSON_RESPONSE_VERSION = void 0;
const randombytes_1 = __importDefault(require("randombytes"));
const skynet_mysky_utils_1 = require("skynet-mysky-utils");
const tweetnacl_1 = require("tweetnacl");
const crypto_1 = require("../crypto");
const string_1 = require("../utils/string");
const validation_1 = require("../utils/validation");
/**
 * The current response version for encrypted JSON files. Part of the metadata
 * prepended to encrypted data.
 */
exports.ENCRYPTED_JSON_RESPONSE_VERSION = 1;
/**
 * The length of the encryption key.
 */
exports.ENCRYPTION_KEY_LENGTH = 32;
/**
 * The length of the hidden field metadata stored along with encrypted files.
 * Note that this is hidden from the user, but not actually encrypted. It is
 * stored after the nonce.
 */
exports.ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH = 16;
/**
 * The length of the random nonce, prepended to the encrypted bytes. It comes
 * before the unencrypted hidden field metadata.
 */
exports.ENCRYPTION_NONCE_LENGTH = 24;
/**
 * The length of the overhead introduced by encryption.
 */
const ENCRYPTION_OVERHEAD_LENGTH = 16;
/**
 * The length of the share-able path seed.
 */
exports.ENCRYPTION_PATH_SEED_LENGTH = 32;
// Descriptive salt that should not be changed.
const SALT_ENCRYPTED_CHILD = "encrypted filesystem child";
// Descriptive salt that should not be changed.
const SALT_ENCRYPTED_TWEAK = "encrypted filesystem tweak";
// Descriptive salt that should not be changed.
const SALT_ENCRYPTION = "encryption";
/**
 * Decrypts the given bytes as an encrypted JSON file.
 *
 * @param data - The given raw bytes.
 * @param key - The encryption key.
 * @returns - The JSON data and metadata.
 * @throws - Will throw if the bytes could not be decrypted.
 */
function decryptJSONFile(data, key) {
    validation_1.validateUint8Array("data", data, "parameter");
    validation_1.validateUint8ArrayLen("key", key, "parameter", exports.ENCRYPTION_KEY_LENGTH);
    // Validate that the size of the data corresponds to a padded block.
    if (!checkPaddedBlock(data.length)) {
        throw new Error(`Expected parameter 'data' to be padded encrypted data, length was '${data.length}', nearest padded block is '${padFileSize(data.length)}'`);
    }
    // Extract the nonce.
    const nonce = data.slice(0, exports.ENCRYPTION_NONCE_LENGTH);
    data = data.slice(exports.ENCRYPTION_NONCE_LENGTH);
    // Extract the unencrypted hidden field metadata.
    const metadataBytes = data.slice(0, exports.ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH);
    data = data.slice(exports.ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH);
    const metadata = decodeEncryptedFileMetadata(metadataBytes);
    if (metadata.version !== exports.ENCRYPTED_JSON_RESPONSE_VERSION) {
        throw new Error(`Received unrecognized JSON response version '${metadata.version}' in metadata, expected '${exports.ENCRYPTED_JSON_RESPONSE_VERSION}'`);
    }
    // Decrypt the non-nonce part of the data.
    let decryptedBytes = tweetnacl_1.secretbox.open(data, nonce, key);
    if (!decryptedBytes) {
        throw new Error("Could not decrypt given encrypted JSON file");
    }
    // Trim the 0-byte padding off the end of the decrypted bytes. This should never remove real data as
    // properly-formatted JSON must end with '}'.
    let paddingIndex = decryptedBytes.length;
    while (paddingIndex > 0 && decryptedBytes[paddingIndex - 1] === 0) {
        paddingIndex--;
    }
    decryptedBytes = decryptedBytes.slice(0, paddingIndex);
    // Parse the final decrypted message as json.
    return JSON.parse(string_1.uint8ArrayToStringUtf8(decryptedBytes));
}
exports.decryptJSONFile = decryptJSONFile;
/**
 * Encrypts the given JSON data and metadata.
 *
 * @param json - The given JSON data.
 * @param metadata - The given metadata.
 * @param key - The encryption key.
 * @returns - The encrypted data.
 */
function encryptJSONFile(json, metadata, key) {
    validation_1.validateObject("json", json, "parameter");
    validation_1.validateNumber("metadata.version", metadata.version, "parameter");
    validation_1.validateUint8ArrayLen("key", key, "parameter", exports.ENCRYPTION_KEY_LENGTH);
    // Stringify the json and convert to bytes.
    let data = string_1.stringToUint8ArrayUtf8(JSON.stringify(json));
    // Add padding so that the final size will be a padded block. The overhead will be added by encryption and we add the nonce at the end.
    const totalOverhead = ENCRYPTION_OVERHEAD_LENGTH + exports.ENCRYPTION_NONCE_LENGTH + exports.ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH;
    const finalSize = padFileSize(data.length + totalOverhead) - totalOverhead;
    data = new Uint8Array([...data, ...new Uint8Array(finalSize - data.length)]);
    // Generate a random nonce.
    const nonce = new Uint8Array(randombytes_1.default(exports.ENCRYPTION_NONCE_LENGTH));
    // Encrypt the data.
    const encryptedBytes = tweetnacl_1.secretbox(data, nonce, key);
    // Prepend the metadata.
    const metadataBytes = encodeEncryptedFileMetadata(metadata);
    const finalBytes = new Uint8Array([...metadataBytes, ...encryptedBytes]);
    // Prepend the nonce to the final data.
    return new Uint8Array([...nonce, ...finalBytes]);
}
exports.encryptJSONFile = encryptJSONFile;
/**
 * Derives key entropy for the given path seed.
 *
 * @param pathSeed - The given path seed.
 * @returns - The key entropy.
 */
function deriveEncryptedFileKeyEntropy(pathSeed) {
    const bytes = new Uint8Array([...crypto_1.sha512(SALT_ENCRYPTION), ...crypto_1.sha512(pathSeed)]);
    const hashBytes = crypto_1.sha512(bytes);
    // Truncate the hash to the size of an encryption key.
    return hashBytes.slice(0, exports.ENCRYPTION_KEY_LENGTH);
}
exports.deriveEncryptedFileKeyEntropy = deriveEncryptedFileKeyEntropy;
/**
 * Derives the encrypted file tweak for the given path seed.
 *
 * @param pathSeed - the given path seed.
 * @returns - The encrypted file tweak.
 */
function deriveEncryptedFileTweak(pathSeed) {
    let hashBytes = crypto_1.sha512(new Uint8Array([...crypto_1.sha512(SALT_ENCRYPTED_TWEAK), ...crypto_1.sha512(pathSeed)]));
    // Truncate the hash or it will be rejected in skyd.
    hashBytes = hashBytes.slice(0, crypto_1.HASH_LENGTH);
    return string_1.toHexString(hashBytes);
}
exports.deriveEncryptedFileTweak = deriveEncryptedFileTweak;
/**
 * Derives the path seed for the relative path, given the starting path seed and
 * whether it is a directory. The path can be an absolute path if the root seed
 * is given.
 *
 * @param pathSeed - The given starting path seed.
 * @param subPath - The path.
 * @param isDirectory - Whether the path is a directory.
 * @returns - The path seed for the given path.
 */
function deriveEncryptedFileSeed(pathSeed, subPath, isDirectory) {
    validation_1.validateHexString("pathSeed", pathSeed, "parameter");
    validation_1.validateString("subPath", subPath, "parameter");
    validation_1.validateBoolean("isDirectory", isDirectory, "parameter");
    let pathSeedBytes = string_1.hexToUint8Array(pathSeed);
    subPath = skynet_mysky_utils_1.sanitizePath(subPath);
    const names = subPath.split("/");
    names.forEach((name, index) => {
        const directory = index === names.length - 1 ? isDirectory : true;
        const derivationPathObj = {
            pathSeed: pathSeedBytes,
            directory,
            name,
        };
        const derivationPath = hashDerivationPathObject(derivationPathObj);
        const bytes = new Uint8Array([...crypto_1.sha512(SALT_ENCRYPTED_CHILD), ...derivationPath]);
        pathSeedBytes = crypto_1.sha512(bytes);
    });
    // Truncate and hex-encode the final output.
    return string_1.toHexString(pathSeedBytes.slice(0, exports.ENCRYPTION_PATH_SEED_LENGTH));
}
exports.deriveEncryptedFileSeed = deriveEncryptedFileSeed;
/**
 * Hashes the derivation path object.
 *
 * @param obj - The given object containing the derivation path.
 * @returns - The hash.
 */
function hashDerivationPathObject(obj) {
    const bytes = new Uint8Array([...obj.pathSeed, obj.directory ? 1 : 0, ...string_1.stringToUint8ArrayUtf8(obj.name)]);
    return crypto_1.sha512(bytes);
}
/**
 * To prevent analysis that can occur by looking at the sizes of files, all
 * encrypted files will be padded to the nearest "pad block" (after encryption).
 * A pad block is minimally 4 kib in size, is always a power of 2, and is always
 * at least 5% of the size of the file.
 *
 * For example, a 1 kib encrypted file would be padded to 4 kib, a 5 kib file
 * would be padded to 8 kib, and a 105 kib file would be padded to 112 kib.
 * Below is a short table of valid file sizes:
 *
 * ```
 *   4 KiB      8 KiB     12 KiB     16 KiB     20 KiB
 *  24 KiB     28 KiB     32 KiB     36 KiB     40 KiB
 *  44 KiB     48 KiB     52 KiB     56 KiB     60 KiB
 *  64 KiB     68 KiB     72 KiB     76 KiB     80 KiB
 *
 *  88 KiB     96 KiB    104 KiB    112 KiB    120 KiB
 * 128 KiB    136 KiB    144 KiB    152 KiB    160 KiB
 *
 * 176 KiB    192 Kib    208 KiB    224 KiB    240 KiB
 * 256 KiB    272 KiB    288 KiB    304 KiB    320 KiB
 *
 * 352 KiB    ... etc
 * ```
 *
 * Note that the first 20 valid sizes are all a multiple of 4 KiB, the next 10
 * are a multiple of 8 KiB, and each 10 after that the multiple doubles. We use
 * this method of padding files to prevent an adversary from guessing the
 * contents or structure of the file based on its size.
 *
 * @param initialSize - The size of the file.
 * @returns - The final size, padded to a pad block.
 * @throws - Will throw if the size would overflow the JS number type.
 */
function padFileSize(initialSize) {
    const kib = 1 << 10;
    // Only iterate to 53 (the maximum safe power of 2).
    for (let n = 0; n < 53; n++) {
        if (initialSize <= (1 << n) * 80 * kib) {
            const paddingBlock = (1 << n) * 4 * kib;
            let finalSize = initialSize;
            if (finalSize % paddingBlock !== 0) {
                finalSize = initialSize - (initialSize % paddingBlock) + paddingBlock;
            }
            return finalSize;
        }
    }
    // Prevent overflow. Max JS number size is 2^53-1.
    throw new Error("Could not pad file size, overflow detected.");
}
exports.padFileSize = padFileSize;
/**
 * Checks if the given size corresponds to the correct padded block.
 *
 * @param size - The given file size.
 * @returns - Whether the size corresponds to a padded block.
 * @throws - Will throw if the size would overflow the JS number type.
 */
function checkPaddedBlock(size) {
    const kib = 1 << 10;
    // Only iterate to 53 (the maximum safe power of 2).
    for (let n = 0; n < 53; n++) {
        if (size <= (1 << n) * 80 * kib) {
            const paddingBlock = (1 << n) * 4 * kib;
            return size % paddingBlock === 0;
        }
    }
    // Prevent overflow. Max JS number size is 2^53-1.
    throw new Error("Could not check padded file size, overflow detected.");
}
exports.checkPaddedBlock = checkPaddedBlock;
/**
 * Decodes the encoded encrypted file metadata.
 *
 * @param bytes - The encoded metadata.
 * @returns - The decoded metadata.
 * @throws - Will throw if the given bytes are of the wrong length.
 */
function decodeEncryptedFileMetadata(bytes) {
    // Ensure input is correct length.
    validation_1.validateUint8ArrayLen("bytes", bytes, "encrypted file metadata bytes", exports.ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH);
    const version = bytes[0];
    return {
        version,
    };
}
/**
 * Encodes the given encrypted file metadata.
 *
 * @param metadata - The given metadata.
 * @returns - The encoded metadata bytes.
 * @throws - Will throw if the version does not fit in a byte.
 */
function encodeEncryptedFileMetadata(metadata) {
    const bytes = new Uint8Array(exports.ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH);
    // Encode the version
    if (metadata.version >= 1 << 8 || metadata.version < 0) {
        throw new Error(`Metadata version '${metadata.version}' could not be stored in a uint8`);
    }
    // Don't need to use a DataView or worry about endianness for a uint8.
    bytes[0] = metadata.version;
    return bytes;
}
exports.encodeEncryptedFileMetadata = encodeEncryptedFileMetadata;
