// skynet-js/dist/cjs/mysky/index.js
"use strict";
/* istanbul ignore file: Much of this functionality is only testable from a browser */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MySky = exports.loadMySky = exports.MAX_ENTRY_LENGTH = exports.MYSKY_ALPHA_DOMAIN = exports.mySkyDevDomain = exports.MYSKY_DEV_DOMAIN = exports.mySkyDomain = exports.MYSKY_DOMAIN = exports.DacLibrary = void 0;
var dac_1 = require("./dac");
Object.defineProperty(exports, "DacLibrary", { enumerable: true, get: function () { return dac_1.DacLibrary; } });
const post_me_1 = require("post-me");
const skynet_mysky_utils_1 = require("skynet-mysky-utils");
const connector_1 = require("./connector");
const client_1 = require("../client");
const utils_1 = require("./utils");
const validation_1 = require("../utils/validation");
const skydb_v2_1 = require("./skydb_v2");
// These imports are deprecated but they are needed for the v1 MySky SkyDB
// methods, which we are keeping so as not to break compatibility.
const skydb_1 = require("./skydb");
/**
 * The domain for MySky.
 */
exports.MYSKY_DOMAIN = "skynet-mysky.hns";
/**
 * @deprecated please use MYSKY_DOMAIN.
 */
exports.mySkyDomain = exports.MYSKY_DOMAIN;
/**
 * The domain for MySky dev.
 */
exports.MYSKY_DEV_DOMAIN = "skynet-mysky-dev.hns";
/**
 * @deprecated please use MYSKY_DEV_DOMAIN.
 */
exports.mySkyDevDomain = exports.MYSKY_DEV_DOMAIN;
/**
 * The domain for MySky alpha. Intentionally not exported in index file.
 */
exports.MYSKY_ALPHA_DOMAIN = "sandbridge.hns";
/**
 * The maximum length for entry data when setting entry data.
 */
exports.MAX_ENTRY_LENGTH = 70;
const mySkyUiRelativeUrl = "ui.html";
const mySkyUiTitle = "MySky UI";
const [mySkyUiW, mySkyUiH] = [640, 750];
/**
 * Loads MySky. Note that this does not log in the user.
 *
 * @param this - The Skynet client.
 * @param skappDomain - The domain of the host skapp. For this domain permissions will be requested and, by default, automatically granted.
 * @param [customOptions] - Additional settings that can optionally be set.
 * @returns - Loaded (but not logged-in) MySky instance.
 */
async function loadMySky(skappDomain, customOptions) {
    const mySky = await MySky.New(this, skappDomain, customOptions);
    return mySky;
}
exports.loadMySky = loadMySky;
/**
 * The singleton object that allows skapp developers to initialize and
 * communicate with MySky.
 */
class MySky {
    // ============
    // Constructors
    // ============
    /**
     * Creates a `MySky` instance.
     *
     * @param connector - The `Connector` object.
     * @param permissions - The initial requested permissions.
     * @param hostDomain - The domain of the host skapp.
     * @param currentPortalUrl - The URL of the current portal. This is the portal that the skapp is running on, not the portal that may have been requested by the developer when creating a `SkynetClient`.
     */
    constructor(connector, permissions, hostDomain, currentPortalUrl) {
        this.connector = connector;
        this.hostDomain = hostDomain;
        this.currentPortalUrl = currentPortalUrl;
        // Holds the loaded DACs.
        this.dacs = [];
        // Holds the currently granted permissions.
        this.grantedPermissions = [];
        // =============
        // SkyDB methods
        // =============
        // v1 (deprecated)
        this.getJSON = skydb_1.getJSON;
        this.getEntryLink = skydb_1.getEntryLink;
        this.setJSON = skydb_1.setJSON;
        this.deleteJSON = skydb_1.deleteJSON;
        this.setDataLink = skydb_1.setDataLink;
        this.getEntryData = skydb_1.getEntryData;
        this.setEntryData = skydb_1.setEntryData;
        this.deleteEntryData = skydb_1.deleteEntryData;
        this.getJSONEncrypted = skydb_1.getJSONEncrypted;
        this.setJSONEncrypted = skydb_1.setJSONEncrypted;
        // v2
        this.dbV2 = {
            getJSON: skydb_v2_1.getJSON.bind(this),
            getEntryLink: skydb_v2_1.getEntryLink.bind(this),
            setJSON: skydb_v2_1.setJSON.bind(this),
            deleteJSON: skydb_v2_1.deleteJSON.bind(this),
            setDataLink: skydb_v2_1.setDataLink.bind(this),
            getEntryData: skydb_v2_1.getEntryData.bind(this),
            setEntryData: skydb_v2_1.setEntryData.bind(this),
            deleteEntryData: skydb_v2_1.deleteEntryData.bind(this),
            getJSONEncrypted: skydb_v2_1.getJSONEncrypted.bind(this),
            setJSONEncrypted: skydb_v2_1.setJSONEncrypted.bind(this),
        };
        if (MySky.instance) {
            throw new Error("Trying to create a second MySky instance");
        }
        this.pendingPermissions = permissions;
    }
    /**
     * Initializes MySky and returns a `MySky` instance.
     *
     * @param client - The Skynet Client.
     * @param [skappDomain] - The domain of the host skapp.
     * @param [customOptions] - Additional settings that can optionally be set.
     * @returns - A `MySky` instance.
     */
    static async New(client, skappDomain, customOptions) {
        const opts = { ...connector_1.DEFAULT_CONNECTOR_OPTIONS, ...customOptions };
        // Enforce singleton.
        if (MySky.instance) {
            return MySky.instance;
        }
        let domain = exports.MYSKY_DOMAIN;
        if (opts.alpha) {
            domain = exports.MYSKY_ALPHA_DOMAIN;
        }
        else if (opts.dev) {
            domain = exports.MYSKY_DEV_DOMAIN;
        }
        const connector = await connector_1.Connector.init(client, domain, customOptions);
        let currentPortalUrl;
        let hostDomain;
        if (window.location.hostname === "localhost") {
            currentPortalUrl = window.location.href;
            hostDomain = "localhost";
        }
        else {
            // MySky expects to be on the same portal as the skapp, so create a new
            // client on the current skapp URL, in case the client the developer
            // instantiated does not correspond to the portal of the current URL.
            const currentUrlClient = new client_1.SkynetClient(window.location.hostname);
            // Trigger a resolve of the portal URL manually. `new SkynetClient` assumes
            // a portal URL is given to it, so it doesn't make the request for the
            // actual portal URL.
            //
            // TODO: We should rework this so it is possible without protected methods.
            //
            // @ts-expect-error - Using protected fields.
            currentUrlClient.customPortalUrl = await currentUrlClient.resolvePortalUrl();
            currentPortalUrl = await currentUrlClient.portalUrl();
            // Get the host domain.
            hostDomain = await currentUrlClient.extractDomain(window.location.hostname);
        }
        // Extract the skapp domain.
        const permissions = [];
        if (skappDomain) {
            const perm1 = new skynet_mysky_utils_1.Permission(hostDomain, skappDomain, skynet_mysky_utils_1.PermCategory.Discoverable, skynet_mysky_utils_1.PermType.Write);
            const perm2 = new skynet_mysky_utils_1.Permission(hostDomain, skappDomain, skynet_mysky_utils_1.PermCategory.Hidden, skynet_mysky_utils_1.PermType.Read);
            const perm3 = new skynet_mysky_utils_1.Permission(hostDomain, skappDomain, skynet_mysky_utils_1.PermCategory.Hidden, skynet_mysky_utils_1.PermType.Write);
            permissions.push(perm1, perm2, perm3);
        }
        MySky.instance = new MySky(connector, permissions, hostDomain, currentPortalUrl);
        // Redirect if we're not on the preferred portal. See
        // `redirectIfNotOnPreferredPortal` for full load flow.
        //
        // TODO: Uncomment the below line once autologin is released. Otherwise the
        // dev console will be spammed with unactionable warnings.
        // try {
        //   await MySky.instance.redirectIfNotOnPreferredPortal();
        // } catch (e) {
        //   // Don't throw an error if we couldn't redirect. The user will never be
        //   // able to log in to MySky and change his preferred portal if MySky can't
        //   // load.
        //   //
        //   // TODO: Add some infrastructure to return warnings to the skapp instead
        //   // of errors?
        //   console.warn(e);
        // }
        return MySky.instance;
    }
    // ==========
    // Public API
    // ==========
    /**
     * Checks if the current browser is supported by MySky.
     *
     * @returns - A promise with a boolean indicating whether the browser is supported and, if not, a string containing the user-friendly error message.
     */
    static async isBrowserSupported() {
        const isSafari = /^((?!chrome|android).)*safari/i.test(navigator.userAgent);
        if (isSafari) {
            return [false, "MySky is currently not supported in Safari browsers."];
        }
        return [true, ""];
    }
    /**
     * Loads the given DACs.
     *
     * @param dacs - The DAC library instances to call `init` on.
     */
    async loadDacs(...dacs) {
        const promises = [];
        for (const dac of dacs) {
            promises.push(this.loadDac(dac));
        }
        this.dacs.push(...dacs);
        await Promise.all(promises);
    }
    /**
     * Adds the given permissions to the list of pending permissions.
     *
     * @param permissions - The list of permissions to add.
     */
    async addPermissions(...permissions) {
        this.pendingPermissions.push(...permissions);
    }
    /**
     * Checks whether main MySky, living in an invisible iframe, is already logged
     * in and all requested permissions are granted.
     *
     * @returns - A boolean indicating whether the user is logged in and all
     * permissions are granted.
     */
    async checkLogin() {
        const [seedFound, permissionsResponse] = await this.connector.connection
            .remoteHandle()
            .call("checkLogin", this.pendingPermissions);
        // Save granted and failed permissions.
        const { grantedPermissions, failedPermissions } = permissionsResponse;
        this.grantedPermissions = grantedPermissions;
        this.pendingPermissions = failedPermissions;
        const loggedIn = seedFound && failedPermissions.length === 0;
        if (loggedIn) {
            await this.handleLogin();
        }
        return loggedIn;
    }
    /**
     * Destroys the mysky connection by:
     *
     * 1. Destroying the connected DACs.
     *
     * 2. Closing the connection.
     *
     * 3. Closing the child iframe.
     *
     * @throws - Will throw if there is an unexpected DOM error.
     */
    async destroy() {
        // TODO: Make sure we are logged out first?
        // TODO: For all connected dacs, send a destroy call.
        // TODO: Delete all connected dacs.
        // Close the connection.
        this.connector.connection.close();
        // Close the child iframe.
        const frame = this.connector.childFrame;
        if (frame) {
            // The parent node should always exist. Sanity check + make TS happy.
            if (!frame.parentNode) {
                throw new Error("'childFrame.parentNode' was not set");
            }
            frame.parentNode.removeChild(frame);
        }
    }
    // TODO: Document what this does exactly.
    /**
     * Log out the user.
     *
     * @returns - An empty promise.
     */
    async logout() {
        await this.connector.connection.remoteHandle().call("logout");
        // Remove auto-relogin if it's set.
        this.connector.client.customOptions.loginFn = undefined;
    }
    /**
     * Requests login access by opening the MySky UI window.
     *
     * @returns - A boolean indicating whether we successfully logged in and all
     * requested permissions were granted.
     */
    async requestLoginAccess() {
        let uiWindow;
        let uiConnection;
        let seedFound = false;
        // Add error listener.
        const { promise: promiseError, controller: controllerError } = (0, skynet_mysky_utils_1.monitorWindowError)();
        // eslint-disable-next-line no-async-promise-executor
        const promise = new Promise(async (resolve, reject) => {
            // Make this promise run in the background and reject on window close or any errors.
            promiseError.catch((err) => {
                if (err === skynet_mysky_utils_1.errorWindowClosed) {
                    // Resolve without updating the pending permissions.
                    resolve();
                    return;
                }
                reject(err);
            });
            try {
                // Launch and connect the UI.
                uiWindow = this.launchUI();
                uiConnection = await this.connectUi(uiWindow);
                // Send the UI the list of required permissions.
                //
                // TODO: This should be a dual-promise that also calls ping() on an interval and rejects if no response was found in a given amount of time.
                const [seedFoundResponse, permissionsResponse] = await uiConnection
                    .remoteHandle()
                    .call("requestLoginAccess", this.pendingPermissions);
                seedFound = seedFoundResponse;
                // Save failed permissions.
                const { grantedPermissions, failedPermissions } = permissionsResponse;
                this.grantedPermissions = grantedPermissions;
                this.pendingPermissions = failedPermissions;
                resolve();
            }
            catch (err) {
                reject(err);
            }
        });
        await promise
            .catch((err) => {
            throw err;
        })
            .finally(() => {
            // Close the window.
            if (uiWindow) {
                uiWindow.close();
            }
            // Close the connection.
            if (uiConnection) {
                uiConnection.close();
            }
            // Clean up the event listeners and promises.
            controllerError.cleanup();
        });
        const loggedIn = seedFound && this.pendingPermissions.length === 0;
        if (loggedIn) {
            await this.handleLogin();
        }
        return loggedIn;
    }
    /**
     * Returns the user ID (i.e. same as the user's public key).
     *
     * @returns - The hex-encoded user ID.
     */
    async userID() {
        return await this.connector.connection.remoteHandle().call("userID");
    }
    /**
     * Lets you get the share-able path seed, which can be passed to
     * `file.getJSONEncrypted` (in file.ts). Requires Hidden Read permission on
     * the path.
     *
     * @param path - The given path.
     * @param isDirectory - Whether the path is a directory.
     * @returns - The seed for the path.
     * @throws - Will throw if the user does not have Hidden Read permission on the path.
     * @deprecated - This function has been deprecated in favor of `getEncryptedPathSeed`.
     */
    async getEncryptedFileSeed(path, isDirectory) {
        return await this.getEncryptedPathSeed(path, isDirectory);
    }
    /**
     * Lets you get the share-able path seed, which can be passed to
     * `file.getJSONEncrypted` (in file.ts). Requires Hidden Read permission on
     * the path.
     *
     * @param path - The given path.
     * @param isDirectory - Whether the path is a directory.
     * @returns - The seed for the path.
     * @throws - Will throw if the user does not have Hidden Read permission on the path.
     */
    async getEncryptedPathSeed(path, isDirectory) {
        (0, validation_1.validateString)("path", path, "parameter");
        (0, validation_1.validateBoolean)("isDirectory", isDirectory, "parameter");
        return await this.connector.connection.remoteHandle().call("getEncryptedFileSeed", path, isDirectory);
    }
    /**
     * signMessage will sign the given data using the MySky user's private key,
     * this method can be used for MySky user verification as the signature may be
     * verified against the user's public key, which is the MySky user id.
     *
     * @param message - message to sign
     * @returns signature
     */
    async signMessage(message) {
        return await this.connector.connection.remoteHandle().call("signMessage", message);
    }
    /**
     * verifyMessageSignature verifies the signature for the message and given
     * public key and returns a boolean that indicates whether the verification
     * succeeded.
     *
     * @param message - the original message that was signed
     * @param signature - the signature
     * @param publicKey - the public key
     * @returns boolean that indicates whether the verification succeeded
     */
    async verifyMessageSignature(message, signature, publicKey) {
        return await this.connector.connection.remoteHandle().call("verifyMessageSignature", message, signature, publicKey);
    }
    // ================
    // Internal Methods
    // ================
    /**
     * Catches any errors returned from the UI and dispatches them in the current
     * window. This is how we bubble up errors from the MySky UI window to the
     * skapp.
     *
     * @param errorMsg - The error message.
     */
    async catchError(errorMsg) {
        const event = new CustomEvent(skynet_mysky_utils_1.dispatchedErrorEvent, { detail: errorMsg });
        window.dispatchEvent(event);
    }
    /**
     * Checks if the MySky user can be logged into a portal account.
     *
     * @returns - Whether the user can be logged into a portal account.
     */
    async checkPortalLogin() {
        return await this.connector.connection.remoteHandle().call("checkPortalLogin");
    }
    /**
     * Launches the MySky UI popup window.
     *
     * @returns - The window handle.
     * @throws - Will throw if the window could not be opened.
     */
    launchUI() {
        const mySkyUrl = new URL(this.connector.url);
        mySkyUrl.pathname = mySkyUiRelativeUrl;
        const uiUrl = mySkyUrl.toString();
        // Open the window.
        const childWindow = (0, utils_1.popupCenter)(uiUrl, mySkyUiTitle, mySkyUiW, mySkyUiH);
        if (!childWindow) {
            throw new Error(`Could not open window at '${uiUrl}'`);
        }
        return childWindow;
    }
    /**
     * Connects to the MySky UI window by establishing a postmessage handshake.
     *
     * @param childWindow - The MySky UI window.
     * @returns - The `Connection` with the other window.
     */
    async connectUi(childWindow) {
        const options = this.connector.options;
        // Complete handshake with UI window.
        const messenger = new post_me_1.WindowMessenger({
            localWindow: window,
            remoteWindow: childWindow,
            remoteOrigin: "*",
        });
        const methods = {
            catchError: this.catchError,
        };
        const connection = await (0, post_me_1.ParentHandshake)(messenger, methods, options.handshakeMaxAttempts, options.handshakeAttemptsInterval);
        return connection;
    }
    /**
     * Gets the preferred portal from MySky, or `null` if not set.
     *
     * @returns - The preferred portal if set.
     */
    async getPreferredPortal() {
        return await this.connector.connection.remoteHandle().call("getPreferredPortal");
    }
    /**
     * Loads the given DAC.
     *
     * @param dac - The dac to load.
     */
    async loadDac(dac) {
        // Initialize DAC.
        await dac.init(this.connector.client, this.connector.options);
        // Add DAC permissions.
        const perms = dac.getPermissions();
        await this.addPermissions(...perms);
    }
    /**
     * Handles the after-login logic.
     */
    async handleLogin() {
        // Call the `onUserLogin` hook for all DACs.
        await Promise.allSettled(this.dacs.map(async (dac) => {
            try {
                await dac.onUserLogin();
            }
            catch (error) {
                // Don't throw on error, just print a console warning.
                console.warn(error);
            }
        }));
        // Redirect if we're not on the preferred portal. See
        // `redirectIfNotOnPreferredPortal` for full login flow.
        //
        // TODO: Uncomment the below line once autologin is released. Otherwise the
        // dev console will be spammed with unactionable warnings.
        // try {
        //   await this.redirectIfNotOnPreferredPortal();
        //   // If we can log in to the portal account, set up auto-relogin.
        //   if (await this.checkPortalLogin()) {
        //     this.connector.client.customOptions.loginFn = this.portalLogin;
        //   }
        // } catch (e) {
        //   // Don't throw an error if we couldn't redirect. The user will never be
        //   // able to log in to MySky and change his preferred portal if MySky can't
        //   // load.
        //   //
        //   // TODO: Add some infrastructure to return warnings to the skapp instead
        //   // of errors?
        //   console.warn(e);
        // }
    }
    /**
     * Logs in to the user's portal account.
     *
     * @returns - An empty promise.
     */
    async portalLogin() {
        return await this.connector.connection.remoteHandle().call("portalLogin");
    }
    /**
     * Get the preferred portal and redirect the page if it is different than
     * the current portal.
     *
     *  Load MySky redirect flow:
     *
     *  1. SDK opens MySky on the same portal as the skapp.
     *  2. If the preferred portal is found in localstorage, MySky connects to it
     *     and we go to step 5.
     *  3. Else, MySky connects to siasky.net.
     *  4. MySky tries to get the saved portal preference.
     *     1. If the portal is set, MySky switches to using the preferred portal.
     *     2. If it is not set or we don't have the seed, MySky switches to using
     *        the current portal as opposed to siasky.net.
     *  5. After MySky finishes loading, SDK queries `mySky.getPortalPreference`.
     *  6. If the preferred portal is set and different than the current portal,
     *     SDK triggers refresh.
     *  7. We go back to step 1 and repeat, but since we're on the right portal
     *     now we won't refresh in step 6.
     *
     * Login redirect flow:
     *
     * 1. SDK logs in through the UI.
     * 2. MySky UI switches to siasky.net and tries to get the saved portal
     *    preference.
     *    1. If the portal is set, MySky switches to using the preferred portal.
     *    2. If it is not set or we don't have the seed, MySky switches to using
     *       the current portal as opposed to siasky.net.
     * 3. SDK queries `mySky.getPortalPreference`.
     * 4. If the preferred portal is set and different than the current portal,
     *    SDK triggers refresh.
     * 5. We go to "Load MySky" step 1 and go through that flow, but we don't
     *    refresh in step 6.
     */
    async redirectIfNotOnPreferredPortal() {
        const currentDomain = window.location.hostname;
        if (currentDomain === "localhost") {
            // Don't redirect on localhost as there is no subdomain to redirect to.
            return;
        }
        // Get the preferred portal.
        const preferredPortalUrl = await this.getPreferredPortal();
        // Is the preferred portal set and different from the current portal?
        if (preferredPortalUrl === null) {
            return;
        }
        else if ((0, utils_1.shouldRedirectToPreferredPortalUrl)(currentDomain, preferredPortalUrl)) {
            // Redirect to the appropriate URL.
            //
            // Get the redirect URL based on the current URL. (Don't use current
            // client as the developer may have set it to e.g. siasky.dev when we are
            // really on siasky.net.)
            const currentDomainClient = new client_1.SkynetClient(currentDomain);
            const newUrl = await (0, utils_1.getRedirectUrlOnPreferredPortal)(currentDomainClient, window.location.hostname, preferredPortalUrl);
            // Check if the portal is valid and working before redirecting.
            const newUrlClient = new client_1.SkynetClient(newUrl);
            const portalUrl = await newUrlClient.portalUrl();
            if (portalUrl) {
                // Redirect.
                redirectPage(newUrl);
            }
        }
        else {
            // If we are on the preferred portal already, we still need to set the
            // client as the developer may have chosen a specific client. We always
            // want to use the user's preference for a portal, if it is set.
            // Set the skapp client to use the user's preferred portal.
            this.connector.client = new client_1.SkynetClient(preferredPortalUrl);
        }
    }
    /**
     * Asks MySky to sign the non-encrypted registry entry.
     *
     * @param entry - The non-encrypted registry entry.
     * @param path - The MySky path.
     * @returns - The signature.
     */
    async signRegistryEntry(entry, path) {
        return await this.connector.connection.remoteHandle().call("signRegistryEntry", entry, path);
    }
    /**
     * Asks MySky to sign the encrypted registry entry.
     *
     * @param entry - The encrypted registry entry.
     * @param path - The MySky path.
     * @returns - The signature.
     */
    async signEncryptedRegistryEntry(entry, path) {
        return await this.connector.connection.remoteHandle().call("signEncryptedRegistryEntry", entry, path);
    }
}
exports.MySky = MySky;
MySky.instance = null;
/**
 * Redirects the page to the given URL.
 *
 * @param url - The URL.
 */
function redirectPage(url) {
    window.location.replace(url);
}
