// skynet-js/dist/cjs/mysky/tweak.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hashPathComponent = exports.splitPath = exports.DiscoverableBucketTweak = exports.deriveDiscoverableFileTweak = void 0;
const crypto_1 = require("../crypto");
const string_1 = require("../utils/string");
const discoverableBucketTweakVersion = 1;
/**
 * Derives the discoverable file tweak for the given path.
 *
 * @param path - The given path.
 * @returns - The hex-encoded tweak.
 */
function deriveDiscoverableFileTweak(path) {
    const dbt = new DiscoverableBucketTweak(path);
    const bytes = dbt.getHash();
    return string_1.toHexString(bytes);
}
exports.deriveDiscoverableFileTweak = deriveDiscoverableFileTweak;
class DiscoverableBucketTweak {
    constructor(path) {
        const paths = splitPath(path);
        const pathHashes = paths.map(hashPathComponent);
        this.version = discoverableBucketTweakVersion;
        this.path = pathHashes;
    }
    encode() {
        const size = 1 + 32 * this.path.length;
        const buf = new Uint8Array(size);
        buf.set([this.version]);
        let offset = 1;
        for (const pathLevel of this.path) {
            buf.set(pathLevel, offset);
            offset += 32;
        }
        return buf;
    }
    getHash() {
        const encoding = this.encode();
        return crypto_1.hashAll(encoding);
    }
}
exports.DiscoverableBucketTweak = DiscoverableBucketTweak;
/**
 * Splits the path by forward slashes.
 *
 * @param path - The path to split.
 * @returns - An array of path components.
 */
function splitPath(path) {
    return path.split("/");
}
exports.splitPath = splitPath;
/**
 * Hashes the path component.
 *
 * @param component - The component extracted from the path.
 * @returns - The hash.
 */
// TODO: Can we replace with hashString?
function hashPathComponent(component) {
    return crypto_1.hashAll(string_1.stringToUint8ArrayUtf8(component));
}
exports.hashPathComponent = hashPathComponent;
