// skynet-js/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "SkynetClient", {
  enumerable: true,
  get: function get() {
    return _client.SkynetClient;
  }
});
Object.defineProperty(exports, "defaultPortalUrl", {
  enumerable: true,
  get: function get() {
    return _utils.defaultPortalUrl;
  }
});
Object.defineProperty(exports, "defaultSkynetPortalUrl", {
  enumerable: true,
  get: function get() {
    return _utils.defaultSkynetPortalUrl;
  }
});
Object.defineProperty(exports, "getRelativeFilePath", {
  enumerable: true,
  get: function get() {
    return _utils.getRelativeFilePath;
  }
});
Object.defineProperty(exports, "getRootDirectory", {
  enumerable: true,
  get: function get() {
    return _utils.getRootDirectory;
  }
});
Object.defineProperty(exports, "parseSkylink", {
  enumerable: true,
  get: function get() {
    return _utils.parseSkylink;
  }
});
Object.defineProperty(exports, "uriHandshakePrefix", {
  enumerable: true,
  get: function get() {
    return _utils.uriHandshakePrefix;
  }
});
Object.defineProperty(exports, "uriHandshakeResolverPrefix", {
  enumerable: true,
  get: function get() {
    return _utils.uriHandshakeResolverPrefix;
  }
});
Object.defineProperty(exports, "uriSkynetPrefix", {
  enumerable: true,
  get: function get() {
    return _utils.uriSkynetPrefix;
  }
});

var _client = require("./client.js");

require("./download.js");

require("./encryption.js");

require("./upload.js");

var _utils = require("./utils.js");
