// skynet-js/dist/mjs/client.js
import axios from "axios";
import { uploadFile, uploadLargeFile, uploadDirectory, uploadDirectoryRequest, uploadSmallFile, uploadSmallFileRequest, uploadLargeFileRequest, } from "./upload";
import { downloadFile, downloadFileHns, getSkylinkUrl, getHnsUrl, getHnsresUrl, getMetadata, getFileContent, getFileContentHns, getFileContentRequest, openFile, openFileHns, resolveHns, } from "./download";
import { getJSONEncrypted, getEntryData, getEntryLink as fileGetEntryLink, getJSON as fileGetJSON } from "./file";
import { pinSkylink } from "./pin";
import { getEntry, getEntryUrl, getEntryLink, setEntry, postSignedEntry } from "./registry";
import { deleteJSON, getJSON, setJSON, setDataLink, getRawBytes } from "./skydb";
import { addUrlQuery, defaultPortalUrl, makeUrl } from "./utils/url";
import { loadMySky } from "./mysky";
import { extractDomain, getFullDomainUrl } from "./mysky/utils";
import { trimSuffix } from "./utils/string";
/**
 * The Skynet Client which can be used to access Skynet.
 */
export class SkynetClient {
    /**
     * The Skynet Client which can be used to access Skynet.
     *
     * @class
     * @param [initialPortalUrl] The initial portal URL to use to access Skynet, if specified. A request will be made to this URL to get the actual portal URL. To use the default portal while passing custom options, pass "".
     * @param [customOptions] Configuration for the client.
     */
    constructor(initialPortalUrl = "", customOptions = {}) {
        // Set methods (defined in other files).
        // Upload
        this.uploadFile = uploadFile;
        this.uploadSmallFile = uploadSmallFile;
        this.uploadSmallFileRequest = uploadSmallFileRequest;
        this.uploadLargeFile = uploadLargeFile;
        this.uploadLargeFileRequest = uploadLargeFileRequest;
        this.uploadDirectory = uploadDirectory;
        this.uploadDirectoryRequest = uploadDirectoryRequest;
        // Download
        this.downloadFile = downloadFile;
        this.downloadFileHns = downloadFileHns;
        this.getSkylinkUrl = getSkylinkUrl;
        this.getHnsUrl = getHnsUrl;
        this.getHnsresUrl = getHnsresUrl;
        this.getMetadata = getMetadata;
        this.getFileContent = getFileContent;
        this.getFileContentHns = getFileContentHns;
        this.getFileContentRequest = getFileContentRequest;
        this.openFile = openFile;
        this.openFileHns = openFileHns;
        this.resolveHns = resolveHns;
        // Pin
        this.pinSkylink = pinSkylink;
        // MySky
        this.extractDomain = extractDomain;
        this.getFullDomainUrl = getFullDomainUrl;
        this.loadMySky = loadMySky;
        // File API
        this.file = {
            getJSON: fileGetJSON.bind(this),
            getEntryData: getEntryData.bind(this),
            getEntryLink: fileGetEntryLink.bind(this),
            getJSONEncrypted: getJSONEncrypted.bind(this),
        };
        // SkyDB
        this.db = {
            deleteJSON: deleteJSON.bind(this),
            getJSON: getJSON.bind(this),
            setJSON: setJSON.bind(this),
            setDataLink: setDataLink.bind(this),
            getRawBytes: getRawBytes.bind(this),
        };
        // SkyDB helpers
        this.registry = {
            getEntry: getEntry.bind(this),
            getEntryUrl: getEntryUrl.bind(this),
            getEntryLink: getEntryLink.bind(this),
            setEntry: setEntry.bind(this),
            postSignedEntry: postSignedEntry.bind(this),
        };
        if (initialPortalUrl === "") {
            // Portal was not given, use the default portal URL. We'll still make a request for the resolved portal URL.
            initialPortalUrl = defaultPortalUrl();
        }
        else {
            // Portal was given, don't make the request for the resolved portal URL.
            this.customPortalUrl = initialPortalUrl;
        }
        this.initialPortalUrl = initialPortalUrl;
        this.customOptions = customOptions;
    }
    /**
     * Make the request for the API portal URL.
     *
     * @returns - A promise that resolves when the request is complete.
     */
    /* istanbul ignore next */
    async initPortalUrl() {
        if (this.customPortalUrl) {
            // Tried to make a request for the API portal URL when a custom URL was already provided.
            return;
        }
        if (!SkynetClient.resolvedPortalUrl) {
            SkynetClient.resolvedPortalUrl = new Promise((resolve, reject) => {
                this.executeRequest({
                    ...this.customOptions,
                    method: "head",
                    url: this.initialPortalUrl,
                    endpointPath: "/",
                }).then((response) => {
                    if (typeof response.headers === "undefined") {
                        reject(new Error("Did not get 'headers' in response despite a successful request. Please try again and report this issue to the devs if it persists."));
                    }
                    const portalUrl = response.headers["skynet-portal-api"];
                    if (!portalUrl) {
                        reject(new Error("Could not get portal URL for the given portal"));
                    }
                    resolve(trimSuffix(portalUrl, "/"));
                });
            });
        }
        await SkynetClient.resolvedPortalUrl;
        return;
    }
    /**
     * Returns the API portal URL. Makes the request to get it if not done so already.
     *
     * @returns - the portal URL.
     */
    /* istanbul ignore next */
    async portalUrl() {
        if (this.customPortalUrl) {
            return this.customPortalUrl;
        }
        // Make the request if needed and not done so.
        this.initPortalUrl();
        return await SkynetClient.resolvedPortalUrl; // eslint-disable-line
    }
    /**
     * Creates and executes a request.
     *
     * @param config - Configuration for the request.
     * @returns - The response from axios.
     */
    async executeRequest(config) {
        const url = await buildRequestUrl(this, config.endpointPath, config.url, config.extraPath, config.query);
        // Build headers.
        const headers = buildRequestHeaders(config.headers, config.customUserAgent, config.customCookie);
        const auth = config.APIKey ? { username: "", password: config.APIKey } : undefined;
        const onUploadProgress = config.onUploadProgress &&
            /* istanbul ignore next */
            function (event) {
                const progress = event.loaded / event.total;
                // Need the if-statement or TS complains.
                if (config.onUploadProgress)
                    config.onUploadProgress(progress, event);
            };
        return axios({
            url,
            method: config.method,
            data: config.data,
            headers,
            auth,
            onUploadProgress,
            responseType: config.responseType,
            transformRequest: config.transformRequest,
            transformResponse: config.transformResponse,
            maxContentLength: Infinity,
            maxBodyLength: Infinity,
            // Allow cross-site cookies.
            withCredentials: true,
        });
    }
}
/**
 * Helper function that builds the request URL.
 *
 * @param client - The Skynet client.
 * @param endpointPath - The endpoint to contact.
 * @param [url] - The base URL to use, instead of the portal URL.
 * @param [extraPath] - An optional path to append to the URL.
 * @param [query] - Optional query parameters to append to the URL.
 * @returns - The built URL.
 */
export async function buildRequestUrl(client, endpointPath, url, extraPath, query) {
    // Build the URL.
    if (!url) {
        const portalUrl = await client.portalUrl();
        url = makeUrl(portalUrl, endpointPath, extraPath !== null && extraPath !== void 0 ? extraPath : "");
    }
    if (query) {
        url = addUrlQuery(url, query);
    }
    return url;
}
/**
 * Helper function that builds the request headers.
 *
 * @param [headers] - Any base headers.
 * @param [customUserAgent] - A custom user agent to set.
 * @param [customCookie] - A custom cookie.
 * @returns - The built headers.
 */
export function buildRequestHeaders(headers, customUserAgent, customCookie) {
    if (!headers) {
        headers = {};
    }
    // Set some headers from common options.
    if (customUserAgent) {
        headers["User-Agent"] = customUserAgent;
    }
    if (customCookie) {
        headers["Cookie"] = customCookie;
    }
    return headers;
}
